src/index.ts

```diff
@@ -0,0 +1,174 @@
+export interface VisualObserverEntry {
+  target: Element;
+  contentRect: DOMRectReadOnly;
+  isAppearing: boolean;
+}
+
+export interface VisualObserverCallback {
+  (this: VisualObserver, entries: VisualObserverEntry[], observer: VisualObserver): void;
+}
+
+interface VisualObserverElement {
+  io: IntersectionObserver | null;
+  threshold: number;
+  isFirstUpdate: boolean;
+}
+
+const floor = Math.floor;
+
+const root = document.documentElement;
+
+/**
+ * Create an observer that notifies when an element is resized, moved, or added/removed from the DOM.
+ */
+export class VisualObserver {
+  #callback: VisualObserverCallback;
+
+  constructor(callback: VisualObserverCallback) {
+    this.#callback = callback;
+  }
+
+  #elements = new Map<Element, VisualObserverElement>();
+
+  #resizeObserver = new ResizeObserver((entries: ResizeObserverEntry[]) => {
+    const visualEntries: VisualObserverEntry[] = [];
+
+    // Any time the root element resizes we need to refresh all the observed elements.
+    if (entries.some((entry) => entry.target === root)) {
+      this.#elements.forEach((_, target) => {
+        // Why force a refresh? we really just need to reset the IntersectionObserver?
```

Comment:
In a typical HTML document, most of the time if the body has been resized then most elements have resized and/or moved. So it’s pretty optimal to refresh each element. But in a spatial canvas where elements are absolutely positioned, this isnt the case and refreshing isn’t necessary. Not sure what to optimize for, but I think if we store the previous DOMRect we can check if the size/position has changed to determine if we need to refresh. A potential configuration to allow

Reply:
Yes, this sounds like a good optimization.

It also occurs to me that `#refreshElement` calls `getClientBoundingRect`, and we may _already have it_ for elements that resize as part of the same tick (i.e., the body causes resize). So this could try to re-use that data too.